synapseclient/models/dataset.py

```diff
@@ -1618,13 +1660,1369 @@ async def snapshot_async(
+        return result
+
+    async def add_item_async(
+        self,
+        item: Union["Dataset", "EntityRef"],
+    ) -> None:
+        """Adds a dataset to the dataset collection.
+        Effect is not seen until the dataset collection is stored.
+
+        Arguments:
+            item: Dataset or EntityRef to add to the collection.
+            synapse_client: The Synapse client to use. Defaults to None.
+
+        Raises:
+            ValueError: If the item is not a Dataset or EntityRef
+
+        Example: Add a Dataset to a Dataset Collection.
+            &nbsp;
+
+            ```python
+            import asyncio
+            from synapseclient import Synapse
+            from synapseclient.models import DatasetCollection, Dataset
+
+            syn = Synapse()
+            syn.login()
+
+            async def main():
+                my_collection = await DatasetCollection(id="syn1234").get_async()
+                await my_collection.add_item_async(Dataset(id="syn1235"))
+                await my_collection.store_async()
+
+            asyncio.run(main())
+            ```
+        """
+        if not isinstance(item, (Dataset, EntityRef)):
+            raise ValueError(
+                f"item must be a Dataset or EntityRef. {item} is a {type(item)}"
+            )
+
+        # EntityRef uses `version`, Dataset uses `version_number`
+        version = item.version if isinstance(item, EntityRef) else item.version_number
+
+        if not any(current_item.id == item.id for current_item in self.items):
+            self.items.append(EntityRef(id=item.id, version=version))
+
+    async def remove_item_async(
+        self,
+        item: Union["Dataset", "EntityRef"],
+    ) -> None:
+        """
+        Removes a dataset from the dataset collection.
+        Effect is not seen until the dataset collection is stored.
+
+        Arguments:
+            item: The Dataset or EntityRef to remove from the collection
+
+        Raises:
+            ValueError: If the item is not a Dataset or EntityRef
+
+        Example: Remove a Dataset from a Dataset Collection.
+            &nbsp;
+
+            ```python
+            import asyncio
+            from synapseclient import Synapse
+            from synapseclient.models import DatasetCollection, Dataset
+
+            syn = Synapse()
+            syn.login()
+
+            async def main():
+                my_collection = await DatasetCollection(id="syn1234").get_async()
+                await my_collection.remove_item_async(Dataset(id="syn1235"))
+                await my_collection.store_async()
+
+            asyncio.run(main())
+            ```
+        """
+        if not isinstance(item, (Dataset, EntityRef)):
+            raise ValueError(
+                f"item must be a Dataset or EntityRef. {item} is a {type(item)}"
+            )
+
+        self.items = [
+            current_item for current_item in self.items if current_item.id != item.id
+        ]
```

Comment:
Should this account for the version of the item we want to remove? I am thinking of a case where someone may have versions `1` and `2` of a single Dataset that they've added to the Dataset Collection. If they call this `remove_item` function intending to remove only version `1`, this would remove both.

Reply:
Good point, I had'nt thought of that case.